internal/handlers/common/projection.go

```diff
@@ -15,160 +15,206 @@
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
 	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
-// ProjectDocuments modifies given documents in places according to the given projection.
-func ProjectDocuments(docs []*types.Document, projection *types.Document) error {
-	if projection.Len() == 0 {
-		return nil
-	}
+var errProjectionEmpty = errors.New("projection is empty")
 
-	inclusion, err := isProjectionInclusion(projection)
-	if err != nil {
-		return err
+// validateProjection check projection document.
+// Document fields could be either included or excluded but not both.
+// Exception is for the _id field that could be included or excluded.
+func validateProjection(projection *types.Document) (*types.Document, bool, error) {
+	validated := types.MakeDocument(0)
+
+	if projection == nil {
+		return nil, false, errProjectionEmpty
 	}
 
-	for i := 0; i < len(docs); i++ {
-		err = projectDocument(inclusion, docs[i], projection)
+	var projectionVal *bool
+
+	iter := projection.Iterator()
+	defer iter.Close()
+
+	for {
+		key, value, err := iter.Next()
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
 		if err != nil {
-			return err
+			return nil, false, lazyerrors.Error(err)
 		}
-	}
 
-	return nil
-}
+		var result bool
 
-// isProjectionInclusion: projection can be only inclusion or exclusion. Validate and return true if inclusion.
-// Exception for the _id field.
-func isProjectionInclusion(projection *types.Document) (inclusion bool, err error) {
-	var exclusion bool
-	for _, k := range projection.Keys() {
-		if k == "_id" { // _id is a special case and can be both
-			continue
-		}
-		v := must.NotFail(projection.Get(k))
-		switch v := v.(type) {
+		switch value := value.(type) {
 		case *types.Document:
-			for _, projectionType := range v.Keys() {
-				err = commonerrors.NewCommandError(
-					commonerrors.ErrNotImplemented,
-					fmt.Errorf("projection of %s is not supported", projectionType),
-				)
-
-				return
-			}
+			return nil, false, commonerrors.NewCommandErrorMsg(
+				commonerrors.ErrNotImplemented,
+				fmt.Sprintf("projection expression %s is not supported", types.FormatAnyValue(value)),
+			)
 
 		case float64, int32, int64:
-			result := types.Compare(v, int32(0))
-			if result == types.Equal {
-				if inclusion {
-					err = commonerrors.NewCommandErrorMsgWithArgument(commonerrors.ErrProjectionExIn,
-						fmt.Sprintf("Cannot do exclusion on field %s in inclusion projection", k),
-						"projection",
-					)
-					return
-				}
-				exclusion = true
-			} else {
-				if exclusion {
-					err = commonerrors.NewCommandErrorMsgWithArgument(commonerrors.ErrProjectionInEx,
-						fmt.Sprintf("Cannot do inclusion on field %s in exclusion projection", k),
-						"projection",
-					)
-					return
-				}
-				inclusion = true
+			comparison := types.Compare(value, int32(0))
+			if comparison != types.Equal {
+				result = true
```

Comment:
So result is true for inclusion. But currently any none zero value (i.e. any negative/positive number) would evaluate projection is inclusion. But I see from tests that only `1` is tested for projection inclusion. Maybe worthwhile adding a test for `-1`?

Reply:
Any non-zero value is treated as `true`, including `-1`.

Reply:
Wow weird,  thanks!

Reply:
Would be great to document this in the code though. Did not find this in the documentation I've looked :see_no_evil:

Reply:
I added a comment to the place where we comparing values.